moveit_core/online_signal_smoothing/src/ruckig_filter.cpp

```diff
@@ -0,0 +1,194 @@
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+#include <moveit/online_signal_smoothing/ruckig_filter.h>
+#include <rclcpp/clock.hpp>
+#include <rclcpp/logging.hpp>
+
+// Disable -Wold-style-cast because all _THROTTLE macros trigger this
+#pragma GCC diagnostic ignored "-Wold-style-cast"
+
+namespace online_signal_smoothing
+{
+namespace
+{
+rclcpp::Logger getLogger()
+{
+  return moveit::getLogger("moveit.core.ruckig_filter_plugin");
+}
+inline constexpr double DEFAULT_JERK_BOUND = 300;  // rad/s^3
+}  // namespace
+
+bool RuckigFilterPlugin::initialize(rclcpp::Node::SharedPtr node, moveit::core::RobotModelConstPtr robot_model,
+                                    size_t num_joints)
+{
+  node_ = node;
+  num_joints_ = num_joints;
+  robot_model_ = robot_model;
+  have_initial_ruckig_output_ = false;
+
+  // get node parameters and store in member variables
+  auto param_listener = online_signal_smoothing::ParamListener(node_);
+  params_ = param_listener.get_params();
+
+  // Ruckig needs the joint vel/accel/jerk bounds.
+  getDefaultJointVelAccelBounds();
+  ruckig::InputParameter<ruckig::DynamicDOFs> ruckig_input(num_joints_);
+  ruckig_input.max_velocity = joint_velocity_bounds_;
+  ruckig_input.max_acceleration = joint_acceleration_bounds_;
+  ruckig_input.max_jerk = joint_jerk_bounds_;
+  ruckig_input.current_position = std::vector<double>(num_joints_, 0.0);
+  ruckig_input.current_velocity = std::vector<double>(num_joints_, 0.0);
+  ruckig_input.current_acceleration = std::vector<double>(num_joints_, 0.0);
+  ruckig_input_ = ruckig_input;
+
+  ruckig_output_.emplace(ruckig::OutputParameter<ruckig::DynamicDOFs>(num_joints_));
+
+  ruckig_.emplace(ruckig::Ruckig<ruckig::DynamicDOFs>(num_joints_, params_.update_period));
+
+  return true;
+}
+
+bool RuckigFilterPlugin::doSmoothing(Eigen::VectorXd& positions, Eigen::VectorXd& velocities,
+                                     Eigen::VectorXd& accelerations)
+{
+  if (have_initial_ruckig_output_ && ruckig_input_ && ruckig_output_)
+  {
+    ruckig_output_->pass_to_input(*ruckig_input_);
+  }
+
+  // Update Ruckig target state
+  // This assumes stationary at the target (zero vel, zero accel)
```

Comment:
This is a big assumption and seems like it may lead to "stop-and-go" behavior. Curious, how does it work if at least the target velocity matches the commanded twist?

Reply:
I thought this was strange too, but when I was developing it several years ago, it's what seemed to work best. It does lead to slower motion overall, though, since Ruckig is planning to stop at every waypoint. Video:

[ruckig_zero_vel.webm](https://github.com/user-attachments/assets/0d63a4cc-5720-418e-b72e-41d4567d834d)


With that said, i'll quickly experiment with non-zero velocities.

Reply:
OK, here's what happens when i pass in the target vels and accels to Ruckig. I was trying to translate along the x-axis only. Pretty crazy.

```
  // Update Ruckig target state
  ruckig_input_->target_position = std::vector<double>(positions.data(), positions.data() + positions.size());
  ruckig_input_->target_velocity = std::vector<double>(velocities.data(), velocities.data() + velocities.size());
  ruckig_input_->target_acceleration = std::vector<double>(accelerations.data(), accelerations.data() + accelerations.size());
  ```
    
[vels_accels.webm](https://github.com/user-attachments/assets/0d1a60eb-9952-4ea9-9146-8fb0941cf5b4)